### DISCUSSION TITLE
Continuous deployment for DAG, airflow.cfg, plugins, and Python dependency from a monorepo?

### DISCUSSION THREAD

**Discussion by jianlongzhong** [apache/airflow#29397] [Q&A]
I’m wondering how do the community deal with Airflow code deployments that involves not only DAG changes, but also changes to plugins, airflow.cfg, and Python dependencies? I'm currently managing a fairly large Airflow deployment with hundreds of active developers who push over 1000 commits per month to a monorepo. While most of the commits are only changing DAG code, there is a good number of commits touching other things like configuration files, plugins, and requirements.txt (which contains all the Python dependencies of our Airflow deployment).

The baseline implementation is we can always shutdown the currently running Airflow cluster, and restart a new one using a Docker image built from the latest code ("shutdown and restart" approach). But this is a rather costly and disruptive process, and due to limitations in our deployment system, it often causes long running jobs in the old cluster to be killed.

Many "Airflow CI/CD" articles online only deal with syncing the code changes of the DAG folder. We implemented a similar mechanism to continuously sync the DAG folder from the latest master branch of our monorepo to the running Airflow cluster. But we quick found two major problems with this kind of approach: (1) whenever there is a change to non DAG code, we have to pause the CD process and do a full redeployment (shutdown the current cluster, rebuild the docker image with latest code, restart the Airflow cluster). (2) In rare cases, simply update the DAG folder with the latest code result in code partially updated. We suspect this is due to tricky issues with Python reloading.

Because of the above situations, we currently have to reverted back to "shutdown and restart" approach whenever we choose to update the production code. If you have experience dealing with similar situations, could you share your experience?

**Comment by hussein-awala** [comment]
For Airflow configuration, Airflow version and  Python dependencies, the best and most secure solution is building Airflow image and restart the pods, and you can do some changes in the docker image to add a graceful shutdown for the worker pods in order to avoid killing the long running tasks (you can check [this blog](https://link.medium.com/G1qTjWNLexb )).

Airflow uses [kubernetes/git-sync](https://github.com/kubernetes/git-sync) to sync the dags files, so you can use the original project to sync the dags files and the plugins, in this case you will not need any restart. Also you can store the git-sync configuration in a configmap and mount it as a volume for all the pods, and sync with git tag instead of master. When you create a new tag, the CD pipeline creates a new config file and patches the configmap, then all the pods will sync with the new tag without any restart.

An alternative solution for dags files and plugins is using a shared volume between the pods, it's a bit expensive but easy to manage where your CD just need to create a new pod to update all the files on this volume.

**Comment by jianlongzhong** [comment]
@hussein-awala Appreciate your inputs! Unfortunately Kubernetes is not available for my Airflow deployment at the moment, so we cannot just restart the pods. Instead, we have to restart the entire deployment (made up of a cluster of EC2s). This is very costly and we want to avoid.

For [kubernetes/git-sync](https://github.com/kubernetes/git-sync), did you experience any issue with refreshing source files while Python is still running (and reading the DAG directory)? We often run into funny issues with that approach, probably due to code inconsistency during file update, or some weird `import` patterns on our user code.

**Reply by potiuk** [reply]
Git - sync provide you atomic "swap" capability so the issues you are having should be gone if you switch to it. This is why I personally hiughly recommend anyone to use git-sync (and ditch shared volumes when you switch to git-sync). You can find more details - including atomic updates -  in the blog post of mine: https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca (in short, git-sync uses symbolic-link swap between old and new versions which solves the "partially updated DAG folder" entirely). 

But yes. if you upgrade plugin code, configuration files, requirements etc. you need to restart Airflow. There is no way around it. But it does not absolutely require you to restart the whole machine, just airflow  - and since you are using anyhow custom deployment on a VM, implementing a script that would monitor for such changes and perform restart (this is especially for plugin code and configuration) should be possible. If you also want to replace installed packages rebuilding docker image (upfront as part of CI verification) is the best idea. Simply build and push your images to registry when needs be in the CI, tag them appropriately, test and when all tests pass - merge the PR. Then write a custom monitoring script that will find out from commit/tag of your Git repo in DAG folder (up to your tagging mechanism) when new image should be used and restart your components with the new image when it is needed (i.e when the expected image to run it has changed). 

Due to the way Git-sync works with symbolic link swaps, you should be even able to keep the old workers running with old images (and reading data from the old "swapped" folder until the running task completes. If you use Celery this can be accomplished by proper signal management:

* Have Celery worker running with Image X
* Build and push image during PR
* Merge the change that requires the new image
* Detect that git-sync update requires new image
* send single (important) SIGTERM to Celery worker - but only to the main process, not the whole process group (it is put in offline mode and running tasks continue running with old code and old image). The celery worker will quit when all tasks are done.
* start a new Celery worker (and other components) using the new image

This will be the least disruptive and "smooth" process to update Airflow.

**Reply by jianlongzhong** [reply]
@potiuk I like your proposed "least disruptive and smooth process to update Airflow". My production environment does not support container management (dynamically start/stop containers) so I'll have to implement something similar but operate on processes instead of containers.

Also, great write up on git-sync! The particular scenario you described was exactly what we experienced (although, not because we use EFS, but due to in place file updates):
> Imagine one of the DAG authors changes the funciton from “shared_function” to “shared_util”. The change is done in both DAG file and “common_code” file and the files got copied to the EFS.
> What happens then ? It depends. If you are in Scheduler, you get an import error. If you are a worker, you have task failure. And both cases are pretty mysterious — because you can locally see that all is correct.

The built-in atomicity of git-sync does look like a game changer for in-place update:
> If your DAG File Processor starts parsing a DAG, it will parse it with a fully consistent version of the DAG and all other files — linked together with the Git Commit they came from.

**Reply by hussein-awala** [reply]
@JianlongZhong if you are not using docker, you can use trigger an [Ansible](https://www.ansible.com/) playbook from your CD pipeline to run commands on your host, and to restart the services, you need to kill the existing processes and re-run them 
```bash
pkill -f -USR2 "airflow scheduler" && airflow scheduler <pass your args> &
pkill -f -USR2 "airflow webserver" && airflow scheduler <pass your args> &
```
